synapse/storage/databases/main/account_data.py

```diff
@@ -390,18 +391,14 @@ async def add_account_data_for_user(
         Returns:
             The maximum stream ID.
         """
-        content_json = json_encoder.encode(content)
-
         async with self._account_data_id_gen.get_next() as next_id:
-            # no need to lock here as account_data has a unique constraint on
-            # (user_id, account_data_type) so simple_upsert will retry if
-            # there is a conflict.
-            await self.db_pool.simple_upsert(
-                desc="add_user_account_data",
-                table="account_data",
-                keyvalues={"user_id": user_id, "account_data_type": account_data_type},
-                values={"stream_id": next_id, "content": content_json},
-                lock=False,
+            await self.db_pool.runInteraction(
+                "add_user_account_data",
+                self._add_account_data_for_user,
+                next_id,
+                user_id,
+                account_data_type,
+                content,
```

Comment:
FYI changing this from an upsert has non-obvious performance cost: `simple_upsert` will be run in `autocommit` mode which a) involves fewer round trips (as no `BEGIN`/`COMMIT` commands) and b) means it doesn't get run in `REPEATABLE READ` isolation level, which has overhead and more likely to conflict with other transactions. I don't think it matters in this case though.

Reply:
It is still doing a `simple_upsert_txn` internally, does that make sense or not really? Hopefully this won't conflict much since it would need to be a user conflicting with themself?

Reply:
It's all correct currently, I just wanted to flag that calling `simple_upsert` is actually not the same as calling `runInteraction` and `simple_upsert_txn` from a perf PoV. I don't think we need to worry about it here though, as we're unlikely to see conflicts (which is where it really matters).